megamolbart/megamolbart/inference_perceiver.py

```diff
@@ -0,0 +1,619 @@
+import torch
+import pandas as pd
+
+import sys
+import numpy as np
+from collections import defaultdict
+import random
+import logging
+from functools import partial
+from typing import List
+import concurrent.futures
+
+from nemo_chem.models.megamolbart.megatron_bart_model import MegaMolBARTModel
+from nemo_chem.models.megamolbart.megatron_bart_latent_model import MegaMolBARTLatentModel
+
+from nemo.collections.nlp.modules.common.megatron.megatron_init import initialize_model_parallel_for_nemo
+
+initialize_model_parallel_for_nemo(
+    world_size=1,
+    global_rank=0,
+    local_rank=0,
+    tensor_model_parallel_size=1,
+    seed=1234,
+)
+logger = logging.getLogger(__name__)
+
+
+class MegaMolBART():
+    def __init__(self, model_dir) -> None:
+        super().__init__()
+
+        torch.set_grad_enabled(False)
+        self.device = 'cuda'
+        self.min_jitter_radius = 1.0
+        self.model, self.version = self.load_model(model_dir)
+        self.max_model_position_embeddings = self.model.max_seq_len
+        self.tokenizer = self.model.tokenizer
+        self.encoder_type = self.model.model.encoder_type
+
+    def _compute_radius(self, scaled_radius):
+        if isinstance(scaled_radius, (int, float)):
+            return float(scaled_radius * self.min_jitter_radius)
+        else:
+            logger.debug(f'Scaled Radius {scaled_radius} Not being used')
+            return self.min_jitter_radius
+
+    def add_jitter(self, embedding, radius, cnt, shape=None):
+        permuted_emb = embedding
+        distorteds = []
+        for _ in range(cnt):
+            noise = torch.normal(0, radius, permuted_emb.shape).to(embedding.device)
+            distorted = (noise + permuted_emb)
+            distorteds.append(distorted)
+
+        return distorteds
+
+    def load_model(self, checkpoint_path):
+        """Load saved model checkpoint
+
+        Params:
+            checkpoint_path: path to nemo checkpoint
+
+        Returns:
+            MegaMolBART trained model
+        """
+        model = MegaMolBARTModel.restore_from(checkpoint_path)
+        model = model.cuda()
+        model.eval()
+        return model, '0.1.0'
+
+    def smiles2embedding(self, smiles, pad_length=None):
+        """Calculate embedding and padding mask for smiles with optional extra padding
+
+        Params
+            smiles: string, input SMILES molecule
+            pad_length: optional extra
+
+        Returns
+            embedding array and boolean mask
+        """
+
+        if isinstance(smiles, str):
+            smiles = [smiles]
+
+        tokens = self.tokenizer.tokenize(smiles, pad=True)
+
+        # # Append to tokens and mask if appropriate. No Padding for Perceiver
+        if pad_length and self.encoder_type != 'perceiver':
+            for i in range(len(tokens['original_tokens'])):
+                n_pad = pad_length - len(tokens['original_tokens'][i])
+                tokens['original_tokens'][i] += [self.tokenizer.pad_token] * n_pad
```

Comment:
This will not raise an error if `n_pad < 0`. Is that the desired behaviour?
Also why not using https://pytorch.org/docs/stable/generated/torch.nn.utils.rnn.pad_sequence.html for padding?